### PR TITLE
Define `Pipelines` and `Components` with simple syntax for common use case of PrimeQA library

### LINKED ISSUE #331: Simplify PrimeQA library consumption for common use cases via pre-defined pipelines
Currently PrimeQA library has lot of moving parts and individual features. For example, Dense Retrievers (ColBERT, DPR), Extractive Reader, Generative Reader, etc.

We need a simplified mechanism to compose these features for common use cases. 
- Extracting relevant text from passages
- Finding relevant documents
- Extracting answers

### PR DESCRIPTION
**Pull request by kpfadnis**
# Pull Request      

## What does this PR do?

```Closes #331 ```       

Notes:
- Replace `(issue)` **above** ↑↑↑ with the issue this PR closes to automatically link the two.
This must be done when the PR is created.
- Add multiple `Closes #(issue)` as needed.
- If this PR is work towards but does not close an issue, simply tag the issue without mentioning `Closes`.

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

- Define `Component` classes for different type of functionalities. 
- Components are combined to form a pipeline based on use case.

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

**Comment by kpfadnis** [comment]
@bhavani105 We can track and comment on the work related issue #331 in this PR. 
@franzmpub We will need to add `DPR` component to complete `dense` retriever components list.

File: primeqa/pipelines/components/indexer/dense.py [hunk 1]
```diff
@@ -107,4 +107,4 @@ def index(self, collection: Union[List[dict], str], *args, **kwargs):
             raise TypeError(
                 "ColBERT indexer expects path to `documents.tsv` as value for `collection` argument."
             )
-        self._indexer.index(self.index_name, collection)
+        self._indexer.index(self.index_name, collection, overwrite='overwrite' in kwargs and kwargs['overwrite'])
```
Comment (kpfadnis) [inline primeqa/pipelines/components/indexer/dense.py:110]: @franzmpub Should we add `overwrite` with default value to datalcass properties for `IndexerComponent`? Or if not common for all indexers maybe on `ColBERTIndexer`? That way we can just take it as part of ColBERTIndexer constructor?

**Review by mihaela-bornea — APPROVED** [review]
LGTM

**Review by franzmpub — APPROVED** [review]
LGTM

**Review by bhavani105 — APPROVED** [review]
LGTM
Minor updates to README.md

**Pull request merged**